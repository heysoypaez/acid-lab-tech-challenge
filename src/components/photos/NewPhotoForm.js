import React from "react";
import "./styles/NewPhotoForm.scss";

function NewPhotoForm(props) {
  return (
    <form className="NewPhotoForm" onSubmit={props.onSubmit}>
      <input
        type="text"
        name="title"
        placeholder="Title"
        onChange={props.onChange}
      />
      <input
        type="url"
        name="url"
        placeholder="Image url"
        onChange={props.onChange}
      />
      <input
        type="url"
        name="thumbnailUrl"
        placeholder="thumbnail url"
        onChange={props.onChange}
      />
      <input type="submit" value="Create New Photo" />
    </form>
  );
}

export default NewPhotoForm;
