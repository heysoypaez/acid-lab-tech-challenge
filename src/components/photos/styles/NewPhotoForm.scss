.NewPhotoForm {
	display: grid;

}
.NewPhotoForm input {
	min-height: 40px;
	padding-left: 1%;
}