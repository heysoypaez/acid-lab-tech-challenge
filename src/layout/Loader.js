import React from "react";
import "./styles/Loader.scss";

function Loader(props) {
  return <section className="Loader">Loading...</section>;
}

export default Loader;
