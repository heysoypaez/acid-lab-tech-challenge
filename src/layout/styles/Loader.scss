.Loader {
	text-align: center;
	padding: 5% 0;
}