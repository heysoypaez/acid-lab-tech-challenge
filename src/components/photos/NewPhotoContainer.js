import React, { Component, Fragment } from "react";
import { graphql } from "@octokit/graphql";
import NewPhotoForm from "./NewPhotoForm.js";
import Loader from "../../layout/Loader.js";
import {Link} from "react-router-dom"

class NewPhotoContainer extends Component {
  state = {
    photos: [],
    loading: false,
    form: {},
    sent: false,
  };

  handleInputChange = (event) => {
    this.setState({
      form: {
        ...this.state.form,
        [event.target.name]: event.target.value,
      },
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    this.createPhoto();
  };

  createPhoto = async () => {
    try {
      const CREATE_PHOTO = `
				mutation (
				  $input: CreatePhotoInput!
				) {
				  createPhoto(input: $input) {
				    title
				    url
				    thumbnailUrl
				  }
				}
			`;

      console.log(this.state.form);

      const { createPhoto } = await graphql({
        url: "https://graphqlzero.almansi.me/api",
        query: CREATE_PHOTO,
        input: {
          title: this.state.form.title,
          url: this.state.form.imgUrl,
          thumbnailUrl: this.state.form.thumbnailUrl,
        },
      });
      console.log(createPhoto);
      this.setState({
        photoSent: createPhoto,
        loading: false,
        error: null,
        sent: true,
      });
    } catch (error) {
      console.log(error.message);
      this.setState({
        error: error,
        loading: false,
      });
    }
  };

  componentDidMount = (props) => {};

  render() {
    if (this.state.loading) {
      return (
        <Fragment>
          <Loader />{" "}
        </Fragment>
      );
    }
    if (this.state.sent) {
      return (
        <Fragment>
          <h2>Photo created Succesfully!</h2>
          <img
            src={this.state.form.imgUrl}
            title={this.state.form.title}
            alt={this.state.form.title}
            width={300}
          />
          <Link to="/photos">See all the photos</Link>
        </Fragment>
      );
    }
    return (
      <Fragment>
        <NewPhotoForm
          onSubmit={this.handleSubmit}
          onChange={this.handleInputChange}
        />
      </Fragment>
    );
  }
}

export default NewPhotoContainer;
