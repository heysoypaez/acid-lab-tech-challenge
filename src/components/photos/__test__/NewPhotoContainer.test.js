import React from "react";
import ReactDOM from "react-dom";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import NewPhotoContainer from "../NewPhotoContainer.js";
import NewPhotoForm from "../NewPhotoForm.js";
import Loader from "../../../layout/Loader.js";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

Enzyme.configure({ adapter: new Adapter() });

describe("Test API fetch", () => {
  test("Data resolved from API should be truthy", async () => {
    const wrapper = shallow(<NewPhotoContainer />);
    const instance = wrapper.instance();
    const input = {
      title: "Photo",
      url:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
      thumbnailUrl:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
    };
    const response = await instance.createPhoto(input);
    expect(response).toBeTruthy();
  });

  test("Data resolved from API should answer an specific object", async () => {
    const wrapper = shallow(<NewPhotoContainer />);
    const instance = wrapper.instance();
    const input = {
      title: "Photo",
      url:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
      thumbnailUrl:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
    };
    const response = await instance.createPhoto(input);
    expect(response).toEqual(input);
  });
});

describe("Test state and props in method", () => {
  test("Initial 'loading' state should be false", () => {
    const wrapper = shallow(<NewPhotoContainer />);
    expect(wrapper.state("loading")).toBe(false);
  });

  test("'Sent' state should be true when createPhoto method is called", async () => {
    const wrapper = shallow(<NewPhotoContainer />);
    const eventSimulated = { preventDefault: () => {} };
    const instance = wrapper.instance();
    const input = {
      title: "Photo",
      url:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
      thumbnailUrl:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
    };
    wrapper.setState({ form: input });
    expect(wrapper.state("sent")).toBe(false);
    await instance.createPhoto(input);
    expect(wrapper.state("sent")).toBe(true);
  });

  test("There should be an error when input is not valid when createPhoto method is called", async () => {
    const wrapper = shallow(<NewPhotoContainer />);
    const eventSimulated = { preventDefault: () => {} };
    const instance = wrapper.instance();
    const input = {
      title: 15159,
      url:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
      thumbnailUrl:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png",
    };
    wrapper.setState({ form: input });
    expect(wrapper.state("sent")).toBeFalsy;
    await instance.createPhoto(input);
    expect(wrapper.state("error")).toBeTruthy;
  });
});

describe("Container contains key components", () => {
  it("Contains a loader component when is loading", () => {
    const wrapper = shallow(<NewPhotoContainer />);
    wrapper.setState({ loading: true });
    expect(wrapper).toContainReact(<Loader />);
  });
});

describe("Snapshot of component", () => {
  it("Matches snapshot in default state", () => {
    const tree = renderer.create(<NewPhotoContainer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
