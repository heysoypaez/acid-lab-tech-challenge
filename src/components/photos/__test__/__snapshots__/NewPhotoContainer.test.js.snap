// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of component Matches snapshot in default state 1`] = `
<form
  className="NewPhotoForm"
  onSubmit={[Function]}
>
  <input
    id="title"
    name="title"
    onChange={[Function]}
    placeholder="Title"
    type="text"
  />
  <input
    id="url"
    name="url"
    onChange={[Function]}
    placeholder="Image url"
    type="url"
  />
  <input
    name="thumbnailUrl"
    onChange={[Function]}
    placeholder="thumbnail url"
    type="url"
  />
  <input
    className="cta"
    type="submit"
    value="Create New Photo"
  />
</form>
`;
